src/WebJobs.Script.WebHost/TokenExpiration/TokenExpirationService.cs

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Diagnostics;
+using Microsoft.Azure.WebJobs.Script.Properties;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.TokenExpiration
+{
+    internal class TokenExpirationService : BackgroundService, IDisposable
+    {
+        private readonly TaskCompletionSource<object> _standby = new();
+        private readonly IEnvironment _environment;
+        private readonly ILogger<TokenExpirationService> _logger;
+        private IDisposable _listener;
+        private CancellationTokenSource _cancellationTokenSource;
+        private Task _analysisTask;
+        private bool _analysisScheduled;
+
+        public TokenExpirationService(IEnvironment environment, ILogger<TokenExpirationService> logger, IOptionsMonitor<StandbyOptions> standbyOptionsMonitor)
+        {
+            _environment = environment;
+            _logger = logger;
+            if (standbyOptionsMonitor.CurrentValue.InStandbyMode)
+            {
+                _listener = standbyOptionsMonitor.OnChange(standbyOptions =>
+                {
+                    if (!standbyOptions.InStandbyMode && !_analysisScheduled)
+                    {
+                        _standby.TrySetResult(null);
+                    }
+                });
+            }
+            else
+            {
+                _standby.TrySetResult(null);
+            }
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+        {
+            await Task.Yield(); // force a yield, so our synchronous code does not slowdown startup
+            await _standby.Task.WaitAsync(cancellationToken);
+
+            if (_environment.IsCoreTools())
+            {
+                return;
+            }
+
+            // Execute first during startup
+            ScheduleTokenExpirationCheck();
```

Comment:
We are already on a background thread here - we can perform the token expiration check directly. No need to use the `_analysisTask` or `Task.Delay` continuations.

``` CSharp
while (true)
{
    AnalyzeSasTokenInUri(); // pass in cancellationToken if you want to sprinkle some .ThrowIfCancellationRequested() calls.
    await Task.Delay(TimeSpan.FromHours(1), cancellationToken);
}
```

Reply:
Should we have the while loop condition be `while(true)` or `while(!cancellationToken.IsCancellationRequested)`?

Reply:
Either is fine. We will exit the loop on cancellation one way or another: either the loop test is false, or the `Task.Delay` throws `OperationCancelledException`.